### PR TITLE
elk-#8 Setup code coverage

### PR DESCRIPTION
**Pull request by elBoberido**
## Notes for Reviewer

This PR adds a CI job to upload the coverage results to codecov.

## Pre-Review Checklist for the PR Author

1. [x] Add sensible notes for the reviewer
1. [x] PR title is short, expressive and meaningful
1. [x] Relevant issues are linked
1. [x] Every source code file has the SPDX header `SPDX-License-Identifier: $(LICENSE_CODE)`
1. [x] Branch follows the naming format (`elk-123-introduce-posix-ipc-example`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Assign PR to reviewer
1. [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/larry-robotics/elkodon/blob/main/doc/release-notes/elkodon-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Unit tests have been written for new behavior
- [x] Public API is documented
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

Use either 'Closes #123' or 'Relates to #123' to reference the corresponding issue.

Relates to #8

### CODE REVIEW

File: scripts/generate-cov-report.sh [hunk 1]
```diff
@@ -11,27 +15,28 @@ LCOV=0
...
     local FILES=$(find . -name "*profraw")
-    llvm-profdata merge -sparse $FILES -o json5format.profdata
+    llvm-profdata merge -sparse $FILES -o ./target/coverage/json5format.profdata
 }
 
 generate() {
     cleanup
     generate_profile
-    merge_report
+    # merge_report
```
Comment (elBoberido) [inline scripts/generate-cov-report.sh:34]: I'm not quite sure what this does and 'show_overview' and 'show_report' did not work on my local install. Can we remove those?

File: scripts/generate-cov-report.sh [hunk 2]
```diff
@@ -54,13 +59,42 @@ show_report() {
...
+          --output-type html \
+          --branch \
+          --ignore-not-existing \
+          --ignore "**/build.rs" \
+          --ignore "**/tests/*" \
+          --ignore "**/examples/*" \
+          --ignore "**/benchmarks/*" \
+          --ignore "**/target/*" \
+          --ignore "**/.cargo/*" \
+          --output-path ./target/coverage/html
```
Comment (elBoberido) [inline scripts/generate-cov-report.sh:77]: I kept the `targe/coverage` path but `cargo clean` does not know about this path and therefore does not clean this folder up. Shall we move everything to `target/debug`?

Comment (elfenpiff) [reply scripts/generate-cov-report.sh:77]: Yes please.

File: scripts/generate-cov-report.sh (hunk 1 above)
Comment (elfenpiff) [reply scripts/generate-cov-report.sh:34]: Can you please replace the `FILES` definition in `show_overview` with

```sh
local FILES=$(find ./target/debug/deps/*tests* -type f -executable)
```

This should repair it. It provides you a nice overview of all files and their code coverage. Not yet perfect but a nice terminal overview.

Comment (elfenpiff) [reply scripts/generate-cov-report.sh:34]: Oh and in `show_report`, please add

```sh
local FILES=$(find ./target/debug/deps/ -type f -executable)
```

at the top and replace the last `CMD` assignment after the for loop with this

```sh
CMD="$CMD --show-instantiation-summary --Xdemangler=rustfilt | less -R"
```

File: scripts/generate-cov-report.sh [hunk 3]
```diff
@@ -42,25 +62,63 @@ show_overview() {
...
+          --ignore-not-existing \
+          --ignore "**/build.rs" \
+          --ignore "**/tests/*" \
+          --ignore "**/examples/*" \
+          --ignore "**/benchmarks/*" \
+          --ignore "**/target/*" \
+          --ignore "**/.cargo/*" \
+          --output-path ./${COVERAGE_DIR}/html
+    sed -i 's/coverage/grcov/' ${COVERAGE_DIR}/html/coverage.json
+    sed -i 's/coverage/grcov/' ${COVERAGE_DIR}/html/badges/*.svg
```
Comment (elBoberido) [inline scripts/generate-cov-report.sh:100]: For some reasons I did this for iceoryx-rs but I've forgotten why :sweat_smile:

**Pull request merged**